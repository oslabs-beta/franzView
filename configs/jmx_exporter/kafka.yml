---
startDelaySeconds: 120
lowercaseOutputName: true
lowercaseOutputLabelNames: true
cacheRules: true

whitelistObjectNames:
  - java.lang:*
  - kafka.cluster:*
  - kafka.controller:*
  - kafka.log:*
  - kafka.server:type=app-info,id=*
  - kafka.server:type=KafkaServer,name=BrokerState
  - kafka.server:type=KafkaRequestHandlerPool,*
  - kafka.server:type=BrokerTopicMetrics,*
  - kafka.server:type=DelayedFetchMetrics,*
  - kafka.server:type=DelayedOperationPurgatory,*
  - kafka.server:type=FetcherLagMetrics,*
  - kafka.server:type=FetcherStats,*
  - kafka.server:type=Request,*
  - kafka.server:type=Fetch,*
  - kafka.server:type=Produce,*
  - kafka.server:type=ReplicaManager,*
  - kafka.server:type=ReplicaFetcherManager,*
  - kafka.server:type=SessionExpireListener,*
  - kafka.server:type=controller-channel-metrics,*
  - kafka.server:type=socket-server-metrics,*
  - kafka.network:type=RequestChannel,*
  - kafka.network:type=Processor,*
  - kafka.network:type=SocketServer,*
  - kafka.network:type=RequestMetrics,*
  - kafka.network:type=RequestMetrics,name=*,request=OffsetCommit,version=*
  - kafka.network:type=RequestMetrics,name=*,request=FetchConsumer,version=*
  - kafka.network:type=RequestMetrics,name=*,request=FetchFollower,version=*
  - kafka.network:type=RequestMetrics,name=*,request=Produce,version=*
  - kafka.network:type=RequestMetrics,name=*,request=Metadata,version=*
  - kafka.coordinator.group:*

rules:
  # This is by far the biggest contributor to the number of sheer metrics being produced.
  # Always keep it on the top for the case of probability when so many metrics will hit the first condition and exit.
  # "kafka.cluster:type=*, name=*, topic=*, partition=*"
  # "kafka.log:type=*,name=*, topic=*, partition=*"
  - pattern: kafka.(\w+)<type=(.+), name=(.+), topic=(.+), partition=(.+)><>Value
    name: kafka_$1_$2_$3
    type: GAUGE
    labels:
      topic: "$4"
      partition: "$5"
  # "kafka.server:type=*,name=*, client-id=*, topic=*, partition=*"
  - pattern: kafka.server<type=(.+), name=(.+), clientId=(.+), topic=(.+), partition=(.*)><>Value
    name: kafka_server_$1_$2
    type: GAUGE
    labels:
      clientId: "$3"
      topic: "$4"
      partition: "$5"
  - pattern: kafka.server<type=(.+), id=(.+)><>StartTimeMs
    name: kafka_server_$1_$2
    type: GAUGE
    labels:
      brokerId: "$2"
  - pattern: kafka.server<type=app-info, name=(.+), clientId=(.+), brokerHost=(.+), brokerPort=(.+)><>Value
    name: kafka_server_$1_$2
    type: GAUGE
    labels:
      clientId: "$3"
      broker: "$4:$5"
  # "kafka.network:type=*, name=*, request=*, error=*"
  # "kafka.network:type=*, name=*, request=*, version=*"
  - pattern: kafka.(\w+)<type=(.+), name=(.+), (.+)=(.+), (.+)=(.+)><>(Count|Value)
    name: kafka_$1_$2_$3
    labels:
      "$4": "$5"
      "$6": "$7"
  - pattern: kafka.(\w+)<type=(.+), name=(.+), (.+)=(.*), (.+)=(.+)><>(\d+)thPercentile
    name: kafka_$1_$2_$3
    type: GAUGE
    labels:
      "$4": "$5"
      "$6": "$7"
      quantile: "0.$8"
  # "kafka.rest:type=*, topic=*, partition=*, client-id=*"
  # "kafka.rest:type=*, cipher=*, protocol=*, client-id=*"
  - pattern: kafka.(\w+)<type=(.+), (.+)=(.+), (.+)=(.+), (.+)=(.+)><>Value
    name: kafka_$1_$2
    labels:
      "$3": "$4"
      "$5": "$6"
      "$7": "$8"
  # Count and Value
  # "kafka.server:type=*, name=*, topic=*"
  # "kafka.server:type=*, name=*, clientId=*"
  # "kafka.server:type=*, name=*, delayedOperation=*"
  # "kafka.server:type=*, name=*, fetcherType=*"
  # "kafka.network:type=*, name=*, networkProcessor=*"
  # "kafka.network:type=*, name=*, processor=*"
  # "kafka.network:type=*, name=*, request=*"
  # "kafka.network:type=*, name=*, listener=*"
  # "kafka.log:type=*, name=*, logDirectory=*"
  # "kafka.log:type=*, name=*, op=*"
  # "kafka.rest:type=*, node-id=*, client-id=*"
  - pattern: kafka.(\w+)<type=(.+), name=(.+), (.+)=(.+)><>(Count|Value)
    name: kafka_$1_$2_$3
    labels:
      "$4": "$5"
  # "kafka.consumer:type=*, topic=*, client-id=*"
  # "kafka.producer:type=*, topic=*, client-id=*"
  # "kafka.rest:type=*, topic=*, client-id=*"
  # "kafka.server:type=*, broker-id=*, fetcher-id=*"
  # "kafka.server:type=*, listener=*, networkProcessor=*"
  - pattern: kafka.(\w+)<type=(.+), (.+)=(.+), (.+)=(.+)><>(Count|Value)
    name: kafka_$1_$2
    labels:
      "$3": "$4"
      "$5": "$6"
  # "kafka.network:type=*, name=*"
  # "kafka.server:type=*, name=*"
  # "kafka.controller:type=*, name=*"
  # "kafka.databalancer:type=*, name=*"
  # "kafka.log:type=*, name=*"
  # "kafka.utils:type=*, name=*"
  - pattern: kafka.(\w+)<type=(.+), name=(.+)><>(Count|Value)
    name: kafka_$1_$2_$3
  # "kafka.producer:type=*, client-id=*"
  # "kafka.producer:type=*, id=*"
  # "kafka.rest:type=*, client-id=*"
  # "kafka.rest:type=*, http-status-code=*"
  # "kafka.server:type=*, BrokerId=*"
  # "kafka.server:type=*, listener=*"
  # "kafka.server:type=*, id=*"
  - pattern: kafka.(\w+)<type=(.+), (.+)=(.+)><>Value
    name: kafka_$1_$2
    labels:
      "$3": "$4"

  - pattern: kafka.server<type=KafkaRequestHandlerPool, name=RequestHandlerAvgIdlePercent><>OneMinuteRate
    name: kafka_server_kafkarequesthandlerpool_requesthandleravgidlepercent_total
    type: GAUGE
  # "kafka.server:type=*, listener=*, networkProcessor=*, clientSoftwareName=*, clientSoftwareVersion=*"
  - pattern: kafka.server<type=socket-server-metrics, clientSoftwareName=(.+), clientSoftwareVersion=(.+), listener=(.+), networkProcessor=(.+)><>connections
    name: kafka_server_socketservermetrics_connections
    type: GAUGE
    labels:
      client_software_name: "$1"
      client_software_version: "$2"
      listener: "$3"
      network_processor: "$4"
  - pattern: "kafka.server<type=socket-server-metrics, listener=(.+), networkProcessor=(.+)><>(.+):"
    name: kafka_server_socketservermetrics_$3
    type: GAUGE
    labels:
      listener: "$1"
      network_processor: "$2"
  # "kafka.coordinator.group:type=*, name=*"
  # "kafka.coordinator.transaction:type=*, name=*"
  - pattern: kafka.coordinator.(\w+)<type=(.+), name=(.+)><>(Count|Value)
    name: kafka_coordinator_$1_$2_$3
  # Percentile
  - pattern: kafka.(\w+)<type=(.+), name=(.+), (.+)=(.*)><>(\d+)thPercentile
    name: kafka_$1_$2_$3
    type: GAUGE
    labels:
      "$4": "$5"
      quantile: "0.$6"
  - pattern: kafka.(\w+)<type=(.+), name=(.+)><>(\d+)thPercentile
    name: kafka_$1_$2_$3
    type: GAUGE
    labels:
      quantile: "0.$4"

  # Quotas
  - pattern : 'kafka.server<type=(Produce|Fetch|Request), user=(.+), client-id=(.+)><>(.+):'
    name: kafka_server_$1_$4
    type: GAUGE
    labels:
      user: "$2"
      client-id: "$3"
  
  - pattern : 'kafka.server<type=(Produce|Fetch|Request), user=(.+)><>(.+):'
    name: kafka_server_$1_$3
    type: GAUGE
    labels:
      user: "$2"
  
  - pattern : 'kafka.server<type=(Produce|Fetch|Request), client-id=(.+)><>(.+):'
    name: kafka_server_$1_$3
    type: GAUGE
    labels:
      client-id: "$2"
