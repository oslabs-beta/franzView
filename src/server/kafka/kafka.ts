import "dotenv/config";
import { Kafka } from "kafkajs";

const brokers = process.env.KAKFA_BROKER.split(",");
const kafka = new Kafka({
  clientId: "franzView-client",
  brokers,
});

const admin = kafka.admin();
const { CONNECT } = admin.events;

admin.on(CONNECT, () => console.log("Kafka Admin Connected!"));

async function run() {
  return await admin.connect();
}

run();

export { admin };
