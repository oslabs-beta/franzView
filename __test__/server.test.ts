import request from "supertest";
import appServer from "../src/server/server";
import { admin } from "../src/server/kafka/kafka";

const server = "http://localhost:3000";

beforeAll(async () => {
  global.testServer = await appServer;
  global.admin = await admin;
});

afterAll(async () => {
  await global.admin.disconnect();
  await global.testServer.stop();
});

describe("REST Server", () => {
  describe("404s for non-existant routes", () => {
    it("Bad POST Request", () => {
      return request(server).post("/badRoute").expect(404);
    });

    it("Bad PUT Request", () => {
      return request(server).put("/badRoute").expect(404);
    });

    it("Bad DELETE Request", () => {
      return request(server).delete("/badRoute").expect(404);
    });
  });
});

describe("GraphQL Queries", () => {
  describe("Cluster Queries", () => {
    it("A query for the cluster type can return the active controller count which is an object with a time field and number.", async () => {
      const result = await global.testServer.executeOperation({
        query: `query Cluster {
          cluster {
            activeControllerCount {
              count: metric
              time
            }
          }
        }`,
      });

      expect(result.errors).toBeUndefined();
      expect(result.data.cluster).toHaveProperty("activeControllerCount");
      expect(result.data.cluster.activeControllerCount).toEqual(
        expect.objectContaining({
          count: expect.any(Number),
          time: expect.any(String),
        })
      );
    });

    it("A query for the cluster type can return the list of brokers in the cluster.", async () => {
      const result = await global.testServer.executeOperation({
        query: `query Cluster {
          cluster {
            brokers {
              brokerHost
              brokerId
              brokerPort
              cpuUsage {
                cpuUsage: metric
                time
              }
              numberUnderReplicatedPartitions {
                  underReplicatedPartitions: metric
                  time
                }
            }
          }
        }`,
      });

      expect(Array.isArray(result.data.cluster.brokers)).toBeTruthy();
      expect(result.data.cluster.brokers).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            brokerId: expect.any(Number),
            brokerPort: expect.any(Number),
            brokerHost: expect.any(String),
            cpuUsage: expect.objectContaining({
              cpuUsage: expect.any(Number),
              time: expect.any(String),
            }),
            numberUnderReplicatedPartitions: expect.objectContaining({
              underReplicatedPartitions: expect.any(Number),
              time: expect.any(String),
            }),
          }),
        ])
      );
    });

    it("A query for the cluster type can return information about which broker is the active controller.", async () => {
      const result = await global.testServer.executeOperation({
        query: `query Cluster {
          cluster {
            brokers {
            brokerHost
            brokerId
            brokerPort
            cpuUsage {
              cpuUsage:metric
              time
            }
            numberUnderReplicatedPartitions {
                underReplicatedPartitions: metric
                time
              }
            }
          }
        }`,
      });

      expect(Array.isArray(result.data.cluster.brokers)).toBeTruthy();
      expect(result.data.cluster.brokers).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            brokerId: expect.any(Number),
            brokerPort: expect.any(Number),
            brokerHost: expect.any(String),
            cpuUsage: expect.objectContaining({
              cpuUsage: expect.any(Number),
              time: expect.any(String),
            }),
            numberUnderReplicatedPartitions: expect.objectContaining({
              underReplicatedPartitions: expect.any(Number),
              time: expect.any(String),
            }),
          }),
        ])
      );
    });

    it("A query for the cluster type can return the offline partition count which is an object with a time field and number.", async () => {
      const result = await global.testServer.executeOperation({
        query: `query Cluster {
          cluster {
            offlinePartitionCount {
                count: metric
                time
              }
            }
          }`,
      });

      expect(result.errors).toBeUndefined();
      expect(result.data.cluster).toHaveProperty("offlinePartitionCount");
      expect(result.data.cluster.offlinePartitionCount).toEqual(
        expect.objectContaining({
          count: expect.any(Number),
          time: expect.any(String),
        })
      );
    });

    it("The cluster type can be queried to return a boolean if a topic can be delete.", async () => {
      const result = await global.testServer.executeOperation({
        query: `query Cluster {
          cluster {
            deleteTopic
            }
          }`,
      });

      expect(result.errors).toBeUndefined();
      expect(typeof result.data.cluster.deleteTopic).toBe("boolean");
    });
  });

  describe("Broker Queries", () => {
    it("A query for a valid broker will have fields: brokerId: Int!, brokerPort: Int!, brokerHost: String!, brokerCpuUsage: BrokerCpuUsage, numberUnderReplicatedPartitions.", async () => {
      const result = await global.testServer.executeOperation({
        query: `query Broker($brokerId: Int!) {
          broker(brokerId: $brokerId) {
              cpuUsage {
                cpuUsage: metric
                time
              }
              numberUnderReplicatedPartitions {
                underReplicatedPartitions: metric
                time
              }
              brokerHost
              brokerPort
              brokerId
            }
          }`,
        variables: { brokerId: 1 },
      });

      expect(result.errors).toBeUndefined();
      expect(typeof result.data.broker.brokerId).toBe("number");
      expect(typeof result.data.broker.brokerHost).toBe("string");
      expect(typeof result.data.broker.brokerPort).toBe("number");
      expect(typeof result.data.broker.cpuUsage.cpuUsage).toBe("number");
      expect(typeof result.data.broker.cpuUsage.time).toBe("string");
      expect(
        typeof result.data.broker.numberUnderReplicatedPartitions
          .underReplicatedPartitions
      ).toBe("number");
      expect(
        typeof result.data.broker.numberUnderReplicatedPartitions.time
      ).toBe("string");
    });

    it("A query for brokers will be an array of brokers", async () => {
      const result = await global.testServer.executeOperation({
        query: `query Brokers {
          brokers {
            brokerHost
            brokerId
            brokerPort
            cpuUsage {
              cpuUsage:metric
              time
            }
            numberUnderReplicatedPartitions {
                underReplicatedPartitions: metric
                time
              }
          }
        }`,
      });

      expect(Array.isArray(result.data.brokers)).toBeTruthy();
      expect(result.data.brokers).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            brokerId: expect.any(Number),
            brokerPort: expect.any(Number),
            brokerHost: expect.any(String),
            cpuUsage: expect.objectContaining({
              cpuUsage: expect.any(Number),
              time: expect.any(String),
            }),
            numberUnderReplicatedPartitions: expect.objectContaining({
              underReplicatedPartitions: expect.any(Number),
              time: expect.any(String),
            }),
          }),
        ])
      );
    });

    it("A query for broker can return a field disk usage which is an object with a time field and number.", async () => {
      const result = await global.testServer.executeOperation({
        query: `query Broker($brokerId: Int!) {
          broker(brokerId: $brokerId) {
              JVMMemoryUsage {
                JVMMemoryUsage: metric
                time
              }
            }
          }`,
        variables: {
          brokerId: 1,
        },
      });
      expect(result.errors).toBeUndefined();
      expect(result.data.broker).toHaveProperty("JVMMemoryUsage");
      expect(result.data.broker.JVMMemoryUsage).toEqual(
        expect.objectContaining({
          JVMMemoryUsage: expect.any(Number),
          time: expect.any(String),
        })
      );
    });
  });
});

describe("GraphQL Mutations", () => {
  describe("Delete Topic", () => {
    beforeEach(async () => {
      await global.testServer.executeOperation({
        query: `mutation AddTopic($name: String!) {
        addTopic(name: $name) {
          name
            }
          }`,
        variables: {
          name: "topicToBeDeleted",
        },
      });
    });

    it("The delete topic mutation returns the topic that was deleted.", async () => {
      const result = await global.testServer.executeOperation({
        query: `mutation DeleteTopic($name: String!) {
          deleteTopic(name: $name) {
            name
          }
        }`,
        variables: { name: "topicToBeDeleted" },
      });

      expect(result.errors).toBeUndefined();
      expect(result).toMatchSnapshot();
    });

    it("Deleting a topic removes it from the cluster and it can no longer be found in the cluster.", async () => {
      await global.testServer.executeOperation({
        query: `mutation DeleteTopic($name: String!) {
          deleteTopic(name: $name) {
            name
          }
        }`,
        variables: { name: "topicToBeDeleted" },
      });

      const response = await global.testServer.executeOperation({
        query: `query topic($name: String!) {
          topic(name: $name) {
            name
          }
        }`,

        variables: { name: "topicToBeDeleted" },
      });

      expect(response.errors).toBeUndefined();
      expect(response.data.topic).toBeNull();
    });
  });

  describe("Add Topic", () => {
    afterEach(async () => {
      await global.testServer.executeOperation({
        query: `mutation DeleteTopic($name: String!) {
        deleteTopic(name: $name) {
          name
            }
          }`,
        variables: {
          name: "newTopic",
        },
      });
    });

    it("The add topic mutation returns the topic that was created.", async () => {
      const result = await global.testServer.executeOperation({
        query: `mutation AddTopic($name: String!) {
          addTopic(name: $name) {
            name
          }
        }`,
        variables: { name: "newTopic" },
      });

      expect(result.errors).toBeUndefined();
      expect(result).toMatchSnapshot();
    });

    it("Adding a topic allows for the topic to be found in the cluster.", async () => {
      const result = await global.testServer.executeOperation({
        query: `mutation AddTopic($name: String!, $replicationFactor: Int, $numPartitions: Int, $configEntries: [ConfigEntry]) {
          addTopic(name: $name, replicationFactor: $replicationFactor, numPartitions: $numPartitions, configEntries: $configEntries) {
            name
            numPartitions
          }
        }`,
        variables: { name: "newTopic" },
      });

      const response = await global.testServer.executeOperation({
        query: `query topic($name: String!) {
          topic(name: $name) {
            name
          }
        }`,

        variables: {
          name: "newTopic",
        },
      });

      expect(result.errors).toBeUndefined();
      expect(response.data.topic.name).toBe(result.data.addTopic.name);
    });
  });
});
