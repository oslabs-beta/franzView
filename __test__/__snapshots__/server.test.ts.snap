// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphQL Mutations Add Topic The add topic mutation returns the topic that was created. 1`] = `
Object {
  "data": Object {
    "addTopic": Object {
      "name": "newTopic",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`GraphQL Mutations Delete Topic The delete topic mutation returns the topic that was deleted. 1`] = `
Object {
  "data": Object {
    "deleteTopic": Object {
      "name": "topicToBeDeleted",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
